WIP: export libpmi2.so.0.0.0 for MPICH derivatives configured --with-pm=slurm --with-pmi=pmi2

garlick opened the pull request:
As discussed in #2347, this PR adds back the PMI-2 API implementation we used to have, in order for our existing mvapich2 configurations to continue to work with the new exec system.   It's effectively a rewrite, and now includes unit tests for the API.   This will need to be deployed along with an MPI "personality" that sets `LD_LIBRARY_PATH` such that our library gets picked up before slurm's by `dlopen()`.

Still todo: test that this actuall works with our mvapich2's  (the whole point!)

garlick commented:
Some fixups squashed, and added commits to set the PMI "kvsname" to the Flux jobid in the shell instead of the string "pmi".  The broker was already fetching the kvsname, so now it doesn't need to fetch the `flux.jobid` key we added, so reverted that code.

chu11 commented on src/common/libpmi/pmi.c:
```diff
@@ -324,7 +324,7 @@ int PMI_Spawn_multiple(int count,
 int PMI_Get_clique_ranks (int ranks[], int length)
 {
-    int result;
+    int result = PMI_ERR_INIT;
```
rebase / merge issue?  It looks like you've added a useless initialization?

garlick replied:
Uh oh.  Thanks!

garlick commented:
Fixed issue reported by @chu11 (and squashed).  Don't know what happened there!

I was able to test on ipa with
```
$  #rm .notce and re-login
$ module load gcc
$ module load mvapich2
$ ./configure && make && make check # note failing t3000-mpi-basic.t
$ flux start -s4
bash-4.2$ ./flux srun -N 4 bash -c "LD_LIBRARY_PATH=../common/.libs:$LD_LIBRARY_PATH ../../t/mpi/hello"
[ipa15:mpi_rank_0][smpi_load_hwloc_topology] WARNING! Invalid my_local_id: -1, Disabling hwloc topology broadcast
0: completed MPI_Init in 0.260s.  There are 4 tasks
0: completed first barrier in 0.000s
0: completed MPI_Finalize in 0.010s
```
so it worked but I need to run down the "Invalid my_local_id" error.

garlick commented:
> [smpi_load_hwloc_topology] WARNING! Invalid my_local_id: -1, Disabling hwloc topology broadcast

It looks like this warning only occurs when we run >1 broker per node.  It's not hard to imagine that hwloc based process binding code could become confused and give up in that case, since Flux doesn't attempt to subdivide resources across brokers.

It's not a problem with one broker per node, multiple nodes, so I think pmi2 is actually working as intended here.

garlick commented:
Just forced a push that adds some minor cleanup tweaks, and drops the client side debugging macros which were a mess, and likely redundant with tools like ltrace(1).   I'll go repeat my mvapich2 test with ltrace once I get to work and make sure we're not giving up any useful capability by dropping this.

garlick commented:
Restarted builder that failed with
```
ERROR: lua/t0001-send-recv.t - exited with status 137 (terminated by signal 9?)
```
and another that failed in valgrind test with:
```
==22143== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
6629flux-start: 0 (pid 22142) Killed
6630not ok 1 - valgrind reports no new errors on 2 broker run
```
I'm guessing slow travis today?

garlick commented:
> I'll go repeat my mvapich2 test with ltrace once I get to work and make sure we're not giving up any useful capability by dropping this.

Yes, in addition to the server side tracing that can be turned on currently by setting the shell verbose flag (will eventually need to be a pmi plugin option, and probably log to a shell logging service), ltrace covers the builtin debugging more effectively that the macros that were in pmi.c.  For example with appropriate ltrace.conf, a 2 node hello compiled with mvapich2 and running under Flux on ipa shows:
```
bash-4.2$ cat ../../ltrace-pmi.conf
int PMI2_Init (+int*, +int*, +int*, +int*);
int PMI2_Finalize ();

int PMI2_Job_GetId (+string, int);
int PMI2_Info_GetJobAttr (string, +string, int, +int*);

int PMI2_KVS_Put (string, string);
int PMI2_KVS_Get (string, int, string, +string, int, +int*);
int PMI2_KVS_Fence ();

bash-4.2$ flux srun -N2 bash -c "LD_LIBRARY_PATH=../common/.libs:$LD_LIBRARY_PATH ltrace -F ../../ltrace-pmi.conf -l 'libpmi2.so*' ../../t/mpi/hello"
libmpi.so.12->PMI2_Init(0, 2, 0, 0)              = 0
libmpi.so.12->PMI2_Job_GetId("22920462073856", 64) = 0
libmpi.so.12->PMI2_Job_GetId("22920462073856", 64) = 0
libmpi.so.12->PMI2_Init(0, 2, 1, 0)              = 0
libmpi.so.12->PMI2_Job_GetId("22920462073856", 64) = 0
libmpi.so.12->PMI2_Info_GetJobAttr("PMI_process_mapping", "(vector,(0,2,1))", 1024, 1) = 0
libmpi.so.12->PMI2_Job_GetId("22920462073856", 64) = 0
libmpi.so.12->PMI2_Info_GetJobAttr("PMI_process_mapping", "(vector,(0,2,1))", 1024, 1) = 0
libmpi.so.12->PMI2_KVS_Put("HOST-1", "-1463810704") = 0
libmpi.so.12->PMI2_KVS_Put("HOST-0", "-1463810960") = 0
libmpi.so.12->PMI2_KVS_Fence()                   = 0
libmpi.so.12->PMI2_KVS_Fence()                   = 0
libmpi.so.12->PMI2_KVS_Get("22920462073856", -1, "HOST-1", "-1463810704", 1024, 11) = 0
libmpi.so.12->PMI2_KVS_Get("22920462073856", -1, "HOST-0", "-1463810960", 1024, 11) = 0
libmpi.so.12->PMI2_KVS_Put("ip0", "63189-74492096") = 0
libmpi.so.12->PMI2_KVS_Put("ip1", "31319-91269312") = 0
libmpi.so.12->PMI2_KVS_Fence()                   = 0
libmpi.so.12->PMI2_KVS_Fence()                   = 0
libmpi.so.12->PMI2_KVS_Get("22920462073856", -1, "ip1", "31319-91269312", 1024, 14) = 0
libmpi.so.12->PMI2_KVS_Get("22920462073856", -1, "ip0", "63189-74492096", 1024, 14) = 0
libmpi.so.12->PMI2_KVS_Fence()                   = 0
libmpi.so.12->PMI2_KVS_Fence()                   = 0
0: completed MPI_Init in 0.687s.  There are 2 tasks
0: completed first barrier in 0.000s
libmpi.so.12->PMI2_KVS_Fence()                   = 0
libmpi.so.12->PMI2_KVS_Fence()                   = 0
libmpi.so.12->PMI2_KVS_Fence()                   = 0
libmpi.so.12->PMI2_KVS_Fence()                   = 0
libmpi.so.12->PMI2_Finalize()                    = 0
0: completed MPI_Finalize in 0.010s
libmpi.so.12->PMI2_Finalize()                    = 0
+++ exited (status 0) +++
+++ exited (status 0) +++
```

grondo commented:
Nice! I poked at this and ran up to 192 tasks with no problems. (largest I could run on fluke)

```
0: completed MPI_Init in 11.770s.  There are 192 tasks
0: completed first barrier in 0.011s
0: completed MPI_Finalize in 0.295s
```

grondo commented:
Restarting a builder that hit a spurious valgrind error, not sure I've seen one like this, but clearly unrelated to this PR:
```
2019-09-19T20:44:09.360796Z broker.crit[0]: child 1 idle for 3 heartbeats
58532019-09-19T20:44:11.370030Z broker.crit[0]: child 1 idle for 4 heartbeats
58542019-09-19T20:44:13.359992Z broker.crit[0]: child 1 idle for 5 heartbeats
58552019-09-19T20:44:15.360229Z broker.crit[0]: child 1 idle for 6 heartbeats
58562019-09-19T20:44:17.360244Z broker.crit[0]: child 1 idle for 7 heartbeats
58572019-09-19T20:44:19.360240Z broker.crit[0]: child 1 idle for 8 heartbeats
5858lt-flux-broker: module 'kvs' was not cleanly shutdown
5859lt-flux-broker: module 'content-sqlite' was not cleanly shutdown
```

garlick commented:
Thanks!
